web/src/reducers/dirty.js

```diff
@@ -0,0 +1,167 @@
+import u from 'updeep';
+
+import {
+  ADD_ACTIVITY_CONTRACTOR,
+  ADD_ACTIVITY_DIRTY,
+  ADD_ACTIVITY_GOAL,
+  ADD_ACTIVITY_EXPENSE,
+  ADD_ACTIVITY_MILESTONE,
+  ADD_ACTIVITY_STATE_PERSON,
+  REMOVE_ACTIVITY,
+  REMOVE_ACTIVITY_CONTRACTOR,
+  REMOVE_ACTIVITY_GOAL,
+  REMOVE_ACTIVITY_EXPENSE,
+  REMOVE_ACTIVITY_MILESTONE,
+  REMOVE_ACTIVITY_STATE_PERSON,
+  UPDATE_ACTIVITY
+} from '../actions/activities';
+import { UPDATE_APD, SAVE_APD_SUCCESS, SELECT_APD } from '../actions/apd';
+
+const initialState = {
+  data: { apd: {}, activities: { byKey: {} } },
+  dirty: false
+};
+
+const mapActivityPropToAPI = prop => {
+  switch (prop) {
+    case 'costAllocationDesc':
+    case 'otherFundingDesc':
+      return 'costAllocationNarrative';
+    default:
+      return prop;
+  }
+};
+
+const up = updates =>
+  typeof updates === 'object'
+    ? Object.entries(updates).reduce(
+        (a, [prop, value]) => ({
+          ...a,
+          [mapActivityPropToAPI(prop)]: up(value)
+        }),
+        {}
+      )
+    : true;
+
+const dirty = (state = initialState, action) => {
+  switch (action.type) {
+    case SAVE_APD_SUCCESS:
+    case SELECT_APD:
+      return initialState;
+
+    case ADD_ACTIVITY_DIRTY:
+      // when an activity is added, we need to add the
+      // whole new activity to the dirty state since it
+      // is populated with initial data
+      return u(
+        {
+          dirty: true,
+          data: { activities: { byKey: { [action.data.key]: action.data } } }
+        },
+        state
+      );
+
+    case ADD_ACTIVITY_CONTRACTOR:
+    case REMOVE_ACTIVITY_CONTRACTOR:
+      return u(
+        {
+          dirty: true,
+          data: {
+            activities: {
+              byKey: { [action.key]: { contractorResources: true } }
```

Comment:
We only need to know that this is dirty - we don't care in what ways.  So just say `true`.  😄